File: app/src/main/java/fr/mpau/tools/UserSettings.java
package fr.mpau.tools;

import android.content.Context;
import android.content.SharedPreferences;
import android.content.SharedPreferences.Editor;
import android.preference.PreferenceManager;
import android.util.Log;

import fr.mpau.models.User;

/**
 * Tools UserSettings
 * <p>
 * - Stockage et récupération du cookie de l'utilisateur depuis les préférences partagées
 * - Stockage et récupération des infos de l'utilisateur depuis les préférences partagées
 * - Stockage et récupération du paramètre signalant le besoin de refresh de la liste des interventions depuis les préférences partagées
 * - Stockage et récupération du paramètre signalant le besoin de restart la liste des interventions depuis les préférences partagées
 * <p>
 * -> setCookie(String cookie, Context context)
 * -> readCookie(Context context)
 * -> deleteCookie(Context context)
 * <p>
 * -> setUser(User user, Context context)
 * -> readUser(Context context)
 * -> deleteUser(Context context)
 * <p>
 * -> setRefreshListInter(Context context)
 * -> readRefreshListInter(Context context)
 * -> deleteRefreshListInter(Context context)
 * <p>
 * -> setRestartListInter(Context context)
 * -> readRestartListInter(Context context)
 * -> deleteRestartListInter(Context context)
 * <p>
 * Author: Jonathan B.
 * Created: 18/11/2017
 * Last Updated: 10/02/2018
 */

public class UserSettings {

    /**
     * Attributs
     */
    private static final String ERROR = "[ERROR]";
    private static final String INFO = "[INFO]";

    /**
     * Enregistre le cookie dans les préférences partagées
     *
     * @param cookie  String
     * @param context Context
     */
    public static void setCookie(String cookie, Context context) {
        // Récupération des préférences
        SharedPreferences preferences = PreferenceManager.getDefaultSharedPreferences(context);
        // Ajout du cookie
        Editor edit = preferences.edit();
        edit.putString("authString", cookie);
        edit.apply();
        Log.e(INFO, "Cookie créé avec succès");
    }

    /**
     * Récupère le cookie dans les préférences partagées
     *
     * @param context Context
     * @return User
     */
    public static String readCookie(Context context) {
        String cookie;
        // Récupération des préférences
        SharedPreferences preferences = PreferenceManager.getDefaultSharedPreferences(context);
        // Récupération du cookie
        cookie = preferences.getString("authString", "");
        return cookie;
    }

    /**
     * Supprime le cookie des préférences partagées
     *
     * @param context Context
     */
    public static void deleteCookie(Context context) {
        // Récupération des préférences
        SharedPreferences preferences = PreferenceManager.getDefaultSharedPreferences(context);
        // Suppression du cookie
        Editor edit = preferences.edit();
        edit.remove("authString");
        edit.apply();
        Log.e(INFO, "Cookie supprimé avec succès");
    }

    /**
     * Enregistre l'utilisateur dans les préférences partagées
     *
     * @param user    User
     * @param context Context
     */
    public static void setUser(User user, Context context) {
        // Récupération des préférences
        SharedPreferences preferences = PreferenceManager.getDefaultSharedPreferences(context);
        // Ajout de l'utilisateur
        Editor edit = preferences.edit();
        edit.putInt("userId", user.getUserId());
        edit.putString("userName", user.getUserName());
        edit.putString("userEmail", user.getUserEmail());
        edit.putInt("userTotalInter", user.getUserTotalInter());
        edit.putInt("userInterIdMax", user.getUserInterIdMax());
        edit.putLong("userInscriptionDate", user.getUserInscriptionDate());
        edit.apply();
        Log.e(INFO, "Utilisateur enregistré avec succès");
    }

    /**
     * Récupère l'utilisateur dans les préférences partagées
     *
     * @param context Context
     * @return User
     */
    public static User readUser(Context context) {
        User user;
        // Récupération des préférences
        SharedPreferences preferences = PreferenceManager.getDefaultSharedPreferences(context);
        // Récupération de l'utilisateur
        int userId = preferences.getInt("userId", 0);
        String userName = preferences.getString("userName", "");
        String userEmail = preferences.getString("userEmail", "");
        int userTotalInter = preferences.getInt("userTotalInter", 0);
        int userInterIdMax = preferences.getInt("userInterIdMax", 0);
        long userInscriptionDate = preferences.getLong("userInscriptionDate", 0);
        if (userId != 0) {
            user = new User(userId, userName, userEmail, userTotalInter, userInterIdMax, userInscriptionDate);
        } else {
            user = null;
            Log.e(ERROR, "Echec de récupération de l'utilisateur");
        }
        return user;
    }

    /**
     * Supprime l'utilisateur des préférences partagées
     *
     * @param context Context
     */
    public static void deleteUser(Context context) {
        // Récupération des préférences
        SharedPreferences preferences = PreferenceManager.getDefaultSharedPreferences(context);
        // Suppression de l'utilisateur
        Editor edit = preferences.edit();
        edit.remove("userId");
        edit.remove("userName");
        edit.remove("userEmail");
        edit.remove("userTotalInter");
        edit.remove("userInterIdMax");
        edit.remove("userInscriptionDate");
        edit.apply();
        Log.e(INFO, "Utilisateur supprimé avec succès");
    }

    /**
     * Enregistre le paramètre 'RefreshListInter' dans les préférences partagées
     *
     * @param context Context
     */
    public static void setRefreshListInter(Context context) {
        // Récupération des préférences
        SharedPreferences preferences = PreferenceManager.getDefaultSharedPreferences(context);
        // Ajout du paramètre
        Editor edit = preferences.edit();
        edit.putBoolean("refreshListInter", true);
        edit.apply();
    }

    /**
     * Récupère le paramètre 'RefreshListInter' dans les préférences partagées
     *
     * @param context Context
     * @return boolean
     */
    public static boolean readRefreshListInter(Context context) {
        boolean refresh;
        // Récupération des préférences
        SharedPreferences preferences = PreferenceManager.getDefaultSharedPreferences(context);
        // Récupération du paramètre
        refresh = preferences.getBoolean("refreshListInter", false);
        return refresh;
    }

    /**
     * Supprime le paramètre 'RefreshListInter' des préférences partagées
     *
     * @param context Context
     */
    public static void deleteRefreshListInter(Context context) {
        // Récupération des préférences
        SharedPreferences preferences = PreferenceManager.getDefaultSharedPreferences(context);
        // Suppression du paramètre
        Editor edit = preferences.edit();
        edit.remove("refreshListInter");
        edit.apply();
    }

    /**
     * Enregistre le paramètre 'RestartListInter' dans les préférences partagées
     *
     * @param context Context
     */
    public static void setRestartListInter(Context context) {
        // Récupération des préférences
        SharedPreferences preferences = PreferenceManager.getDefaultSharedPreferences(context);
        // Ajout du paramètre
        Editor edit = preferences.edit();
        edit.putBoolean("restartListInter", true);
        edit.apply();
    }

    /**
     * Récupère le paramètre 'RestartListInter' dans les préférences partagées
     *
     * @param context Context
     * @return boolean
     */
    public static boolean readRestartListInter(Context context) {
        boolean restart;
        // Récupération des préférences
        SharedPreferences preferences = PreferenceManager.getDefaultSharedPreferences(context);
        // Récupération du paramètre
        restart = preferences.getBoolean("restartListInter", false);
        return restart;
    }

    /**
     * Supprime le paramètre 'RestartListInter' des préférences partagées
     *
     * @param context Context
     */
    public static void deleteRestartListInter(Context context) {
        // Récupération des préférences
        SharedPreferences preferences = PreferenceManager.getDefaultSharedPreferences(context);
        // Suppression du paramètre
        Editor edit = preferences.edit();
        edit.remove("restartListInter");
        edit.apply();
    }

}

File: src/main/java/fr/mpau/tool/UserSettings.kt
package fr.mpau.tool

import android.content.Context
import android.content.SharedPreferences
import android.util.Log
import fr.mpau.bean.User

/**
 * Tool - UserSettings
 *
 * - Stockage et récupération du cookie de l'utilisateur depuis les préférences partagées
 * - Stockage et récupération des infos de l'utilisateur depuis les préférences partagées
 * - Stockage et récupération du paramètre signalant le besoin de refresh de la liste des interventions depuis les préférences partagées
 * - Stockage et récupération du paramètre signalant le besoin de restart la liste des interventions depuis les préférences partagées
 *
 * -> setCookie(String cookie, Context context)
 * -> readCookie(Context context)
 * -> deleteCookie(Context context)
 *
 * -> setUser(User user, Context context)
 * -> readUser(Context context)
 * -> deleteUser(Context context)
 *
 * -> setRefreshListInter(Context context)
 * -> readRefreshListInter(Context context)
 * -> deleteRefreshListInter(Context context)
 *
 * -> setRestartListInter(Context context)
 * -> readRestartListInter(Context context)
 * -> deleteRestartListInter(Context context)
 *
 * Author: Jonathan
 * Created: 18/11/2017
 * Last Updated: 24/01/2025
 */
object UserSettings {

    /**
     * Attributs
     */

    private const val TAG = "UserSettings"
    private const val MPAU_PREFS = "MPAU_Prefs"

    /**
     * Enregistre le cookie dans les préférences partagées
     *
     * @param cookie String
     * @param context Context
     */
    fun setCookie(cookie: String, context: Context) {
        // Récupération des préférences
        val preferences: SharedPreferences =
            context.getSharedPreferences(MPAU_PREFS, Context.MODE_PRIVATE)
        // Ajout du cookie
        val edit: SharedPreferences.Editor = preferences.edit()
        edit.putString("auth_string", cookie)
        edit.apply()
        Log.d(TAG, "Cookie créé avec succès")
    }

    /**
     * Récupère le cookie dans les préférences partagées
     *
     * @param context Context
     * @return String
     */
    fun readCookie(context: Context): String {
        val cookie: String
        // Récupération des préférences
        val preferences: SharedPreferences =
            context.getSharedPreferences(MPAU_PREFS, Context.MODE_PRIVATE)
        // Récupération du cookie
        cookie = preferences.getString("auth_string", "").toString()
        return cookie
    }

    /**
     * Supprime le cookie des préférences partagées
     *
     * @param context Context
     */
    fun deleteCookie(context: Context) {
        // Récupération des préférences
        val preferences: SharedPreferences =
            context.getSharedPreferences(MPAU_PREFS, Context.MODE_PRIVATE)
        // Suppression du cookie
        val edit: SharedPreferences.Editor = preferences.edit()
        edit.remove("auth_string")
        edit.apply()
        Log.d(TAG, "Cookie supprimé avec succès")
    }

    /**
     * Enregistre l'utilisateur dans les préférences partagées
     *
     * @param user User
     * @param context Context
     */
    fun setUser(user: User, context: Context) {
        // Récupération des préférences
        val preferences: SharedPreferences =
            context.getSharedPreferences(MPAU_PREFS, Context.MODE_PRIVATE)
        // Ajout de l'utilisateur
        val edit: SharedPreferences.Editor = preferences.edit()
        edit.putInt("user_id", user.id)
        edit.putString("user_name", user.name)
        edit.putString("user_pass", user.pass)
        edit.putString("user_email", user.email)
        edit.putInt("user_nb_inter", user.nbInter)
        edit.putInt("user_inter_id_max", user.interIdMax)
        edit.putLong("user_inscription_date", user.inscriptionDate)
        edit.apply()
        Log.d(TAG, "Utilisateur enregistré avec succès")
    }

    /**
     * Récupère l'utilisateur dans les préférences partagées
     *
     * @param context Context
     * @return User?
     */
    fun readUser(context: Context): User? {
        val user: User?
        // Récupération des préférences
        val preferences: SharedPreferences =
            context.getSharedPreferences(MPAU_PREFS, Context.MODE_PRIVATE)
        // Récupération de l'utilisateur
        val id: Int = preferences.getInt("user_id", 0)
        val name: String = preferences.getString("user_name", "").toString()
        val pass: String = preferences.getString("user_pass", "").toString()
        val email: String = preferences.getString("user_email", "").toString()
        val nbInter: Int = preferences.getInt("user_nb_inter", 0)
        val interIdMax: Int = preferences.getInt("user_inter_id_max", 0)
        val inscriptionDate: Long = preferences.getLong("user_inscription_date", 0)
        if (id != 0) {
            user = User(id, name, pass, email, nbInter, interIdMax, inscriptionDate)
        } else {
            user = null
            Log.e(TAG, "Echec de récupération de l'utilisateur")
        }
        return user
    }

    /**
     * Supprime l'utilisateur des préférences partagées
     *
     * @param context Context
     */
    fun deleteUser(context: Context) {
        // Récupération des préférences
        val preferences: SharedPreferences =
            context.getSharedPreferences(MPAU_PREFS, Context.MODE_PRIVATE)
        // Suppression de l'utilisateur
        val edit: SharedPreferences.Editor = preferences.edit()
        edit.remove("user_id")
        edit.remove("user_name")
        edit.remove("user_pass")
        edit.remove("user_email")
        edit.remove("user_nb_inter")
        edit.remove("user_inter_id_max")
        edit.remove("user_inscription_date")
        edit.apply()
        Log.d(TAG, "Utilisateur supprimé avec succès")
    }

    /**
     * Enregistre le paramètre 'refresh_list_inter' dans les préférences partagées
     *
     * @param context Context
     */
    fun setRefreshListInter(context: Context) {
        // Récupération des préférences
        val preferences: SharedPreferences =
            context.getSharedPreferences(MPAU_PREFS, Context.MODE_PRIVATE)
        // Ajout du paramètre
        val edit: SharedPreferences.Editor = preferences.edit()
        edit.putBoolean("refresh_list_inter", true)
        edit.apply()
    }

    /**
     * Récupère le paramètre 'refresh_list_inter' dans les préférences partagées
     *
     * @param context Context
     * @return Boolean
     */
    fun readRefreshListInter(context: Context): Boolean {
        val refresh: Boolean
        // Récupération des préférences
        val preferences: SharedPreferences =
            context.getSharedPreferences(MPAU_PREFS, Context.MODE_PRIVATE)
        // Récupération du paramètre
        refresh = preferences.getBoolean("refresh_list_inter", false)
        return refresh
    }

    /**
     * Supprime le paramètre 'refresh_list_inter' des préférences partagées
     *
     * @param context Context
     */
    fun deleteRefreshListInter(context: Context) {
        // Récupération des préférences
        val preferences: SharedPreferences =
            context.getSharedPreferences(MPAU_PREFS, Context.MODE_PRIVATE)
        // Suppression du paramètre
        val edit: SharedPreferences.Editor = preferences.edit()
        edit.remove("refresh_list_inter")
        edit.apply()
    }

    /**
     * Enregistre le paramètre 'restart_list_inter' dans les préférences partagées
     *
     * @param context Context
     */
    fun setRestartListInter(context: Context) {
        // Récupération des préférences
        val preferences: SharedPreferences =
            context.getSharedPreferences(MPAU_PREFS, Context.MODE_PRIVATE)
        // Ajout du paramètre
        val edit: SharedPreferences.Editor = preferences.edit()
        edit.putBoolean("restart_list_inter", true)
        edit.apply()
    }

    /**
     * Récupère le paramètre 'restart_list_inter' dans les préférences partagées
     *
     * @param context Context
     * @return Boolean
     */
    fun readRestartListInter(context: Context): Boolean {
        val restart: Boolean
        // Récupération des préférences
        val preferences: SharedPreferences =
            context.getSharedPreferences(MPAU_PREFS, Context.MODE_PRIVATE)
        // Récupération du paramètre
        restart = preferences.getBoolean("restart_list_inter", false)
        return restart
    }

    /**
     * Supprime le paramètre 'restart_list_inter' des préférences partagées
     *
     * @param context Context
     */
    fun deleteRestartListInter(context: Context) {
        // Récupération des préférences
        val preferences: SharedPreferences =
            context.getSharedPreferences(MPAU_PREFS, Context.MODE_PRIVATE)
        // Suppression du paramètre
        val edit: SharedPreferences.Editor = preferences.edit()
        edit.remove("restart_list_inter")
        edit.apply()
    }

}
